// js-confuser/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Obfuscator", {
  enumerable: true,
  get: function () {
    return _obfuscator.default;
  }
});
Object.defineProperty(exports, "Transform", {
  enumerable: true,
  get: function () {
    return _transform.default;
  }
});
exports.default = exports.debugTransformations = exports.debugObfuscation = void 0;
exports.obfuscate = obfuscate;
exports.obfuscateAST = obfuscateAST;
Object.defineProperty(exports, "presets", {
  enumerable: true,
  get: function () {
    return _presets.default;
  }
});

var _compiler = _interopRequireWildcard(require("./compiler"));

var _parser = _interopRequireWildcard(require("./parser"));

var _obfuscator = _interopRequireDefault(require("./obfuscator"));

var _transform = _interopRequireDefault(require("./transforms/transform"));

var _object = require("./util/object");

var _presets = _interopRequireDefault(require("./presets"));

var assert = _interopRequireWildcard(require("assert"));

var _options = require("./options");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * **JsConfuser**: Obfuscates JavaScript.
 * @param code - The code to be obfuscated.
 * @param options - An object of obfuscation options: `{preset: "medium", target: "browser"}`.
 */
async function obfuscate(code, options) {
  return await JsConfuser(code, options);
}
/**
 * Obfuscates an [ESTree](https://github.com/estree/estree) compliant AST.
 *
 * **Note:** Mutates the object.
 *
 * @param AST - The [ESTree](https://github.com/estree/estree) compliant AST. This object will be mutated.
 * @param options - The obfuscation options.
 *
 * [See all settings here](https://github.com/MichaelXF/js-confuser#options)
 */


async function obfuscateAST(AST, options) {
  assert.ok(typeof AST === "object", "AST must be type object");
  assert.ok(AST.type == "Program", "AST.type must be equal to 'Program'");
  (0, _options.validateOptions)(options);
  options = await (0, _options.correctOptions)(options);
  var obfuscator = new _obfuscator.default(options);
  await obfuscator.apply(AST);
  options.verbose && console.log("* Removing $ properties");
  (0, _object.remove$Properties)(AST);
}
/**
 * **JsConfuser**: Obfuscates JavaScript.
 * @param code - The code to be obfuscated.
 * @param options - An object of obfuscation options: `{preset: "medium", target: "browser"}`.
 */


var JsConfuser = async function (code, options) {
  if (typeof code !== "string") {
    throw new TypeError("code must be type string");
  }

  (0, _options.validateOptions)(options);
  options = await (0, _options.correctOptions)(options);
  options.verbose && console.log("* Parsing source code");
  var tree = await (0, _parser.default)(code);
  options.verbose && console.log("* Obfuscating...");
  var obfuscator = new _obfuscator.default(options);
  await obfuscator.apply(tree);
  options.verbose && console.log("* Removing $ properties");
  (0, _object.remove$Properties)(tree);
  options.verbose && console.log("* Generating code");
  var result = await (0, _compiler.default)(tree, options);
  return result;
};

var debugTransformations = async function debugTransformations(code, options) {
  (0, _options.validateOptions)(options);
  options = await (0, _options.correctOptions)(options);
  var frames = [];
  var tree = (0, _parser.parseSync)(code);
  var obfuscator = new _obfuscator.default(options);
  var time = Date.now();
  obfuscator.on("debug", (name, tree) => {
    frames.push({
      name: name,
      code: (0, _compiler.compileJsSync)(tree, options),
      ms: Date.now() - time
    });
    time = Date.now();
  });
  await obfuscator.apply(tree, true);
  return frames;
};

exports.debugTransformations = debugTransformations;

var debugObfuscation = async function debugTransformations(code, options, callback) {
  (0, _options.validateOptions)(options);
  options = await (0, _options.correctOptions)(options);
  var tree = (0, _parser.parseSync)(code);
  var obfuscator = new _obfuscator.default(options);
  var totalTransforms = obfuscator.array.length;
  obfuscator.on("debug", (name, tree, i) => {
    callback(name, i, totalTransforms);
  });
  await obfuscator.apply(tree, true);
  var output = (0, _compiler.default)(tree, options);
  return output;
};

exports.debugObfuscation = debugObfuscation;
JsConfuser.obfuscate = obfuscate;
JsConfuser.obfuscateAST = obfuscateAST;
JsConfuser.presets = _presets.default;
JsConfuser.debugTransformations = debugTransformations;
JsConfuser.debugObfuscation = debugObfuscation;
JsConfuser.Obfuscator = _obfuscator.default;
JsConfuser.Transform = _transform.default;

if (typeof window !== "undefined") {
  window["JsConfuser"] = JsConfuser;
}

if (typeof global !== "undefined") {
  global["JsConfuser"] = JsConfuser;
}

var _default = JsConfuser;
exports.default = _default;
