// js-confuser/dist/order.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ObfuscateOrder = void 0;

/**
 * Describes the order of transformations.
 */
let ObfuscateOrder;
exports.ObfuscateOrder = ObfuscateOrder;

(function (ObfuscateOrder) {
  ObfuscateOrder[ObfuscateOrder["Preparation"] = 0] = "Preparation";
  ObfuscateOrder[ObfuscateOrder["ObjectExtraction"] = 1] = "ObjectExtraction";
  ObfuscateOrder[ObfuscateOrder["Flatten"] = 2] = "Flatten";
  ObfuscateOrder[ObfuscateOrder["RGF"] = 3] = "RGF";
  ObfuscateOrder[ObfuscateOrder["Lock"] = 4] = "Lock";
  ObfuscateOrder[ObfuscateOrder["Dispatcher"] = 6] = "Dispatcher";
  ObfuscateOrder[ObfuscateOrder["DeadCode"] = 8] = "DeadCode";
  ObfuscateOrder[ObfuscateOrder["Calculator"] = 9] = "Calculator";
  ObfuscateOrder[ObfuscateOrder["ControlFlowFlattening"] = 10] = "ControlFlowFlattening";
  ObfuscateOrder[ObfuscateOrder["Eval"] = 11] = "Eval";
  ObfuscateOrder[ObfuscateOrder["GlobalConcealing"] = 12] = "GlobalConcealing";
  ObfuscateOrder[ObfuscateOrder["OpaquePredicates"] = 13] = "OpaquePredicates";
  ObfuscateOrder[ObfuscateOrder["StringSplitting"] = 16] = "StringSplitting";
  ObfuscateOrder[ObfuscateOrder["StringConcealing"] = 17] = "StringConcealing";
  ObfuscateOrder[ObfuscateOrder["StringCompression"] = 18] = "StringCompression";
  ObfuscateOrder[ObfuscateOrder["HideInitializingCode"] = 19] = "HideInitializingCode";
  ObfuscateOrder[ObfuscateOrder["Stack"] = 20] = "Stack";
  ObfuscateOrder[ObfuscateOrder["DuplicateLiteralsRemoval"] = 22] = "DuplicateLiteralsRemoval";
  ObfuscateOrder[ObfuscateOrder["Shuffle"] = 24] = "Shuffle";
  ObfuscateOrder[ObfuscateOrder["NameRecycling"] = 25] = "NameRecycling";
  ObfuscateOrder[ObfuscateOrder["MovedDeclarations"] = 26] = "MovedDeclarations";
  ObfuscateOrder[ObfuscateOrder["RenameVariables"] = 27] = "RenameVariables";
  ObfuscateOrder[ObfuscateOrder["RenameLabels"] = 28] = "RenameLabels";
  ObfuscateOrder[ObfuscateOrder["Minify"] = 30] = "Minify";
  ObfuscateOrder[ObfuscateOrder["ES5"] = 31] = "ES5";
  ObfuscateOrder[ObfuscateOrder["StringEncoding"] = 32] = "StringEncoding";
  ObfuscateOrder[ObfuscateOrder["AntiTooling"] = 34] = "AntiTooling";
  ObfuscateOrder[ObfuscateOrder["HexadecimalNumbers"] = 35] = "HexadecimalNumbers";
})(ObfuscateOrder || (exports.ObfuscateOrder = ObfuscateOrder = {}));
