// js-confuser/dist/transforms/es5/antiDestructuring.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = require("assert");

var _template = _interopRequireDefault(require("../../templates/template"));

var _traverse = require("../../traverse");

var _gen = require("../../util/gen");

var _insert = require("../../util/insert");

var _transform = _interopRequireDefault(require("../transform"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Removes destructuring from function parameters.
 *
 * ```
 * // input
 * function({property}){
 * }
 *
 * // output
 * function(){
 *  var [{property}] = arguments;
 * }
 *
 * // input
 * var fn = ({property})=>{};
 *
 * // output
 * var fn = (_)=>{
 *  var [{property}] = [_];
 * }
 * ```
 */
class AntiDestructuringParameters extends _transform.default {
  constructor(o) {
    super(o);
  }

  match(object, parents) {
    return (object.param || object.params) && object.body;
  }

  transform(object, parents) {
    return () => {
      if (object.param) {
        // Catch clause
        if (object.param.type != "Identifier") {
          var catchName = this.getPlaceholder();
          var cloned = { ...object.param
          };
          object.param = (0, _gen.Identifier)(catchName);
          (0, _insert.getBlockBody)(object.body).unshift((0, _gen.VariableDeclaration)([(0, _gen.VariableDeclarator)(cloned, (0, _gen.Identifier)(catchName))]));
        }

        return;
      } // For function parameters


      var isDestructed = false;
      var parameters = object.params;
      (0, _traverse.walk)(parameters, [object, ...parents], (o, p) => {
        if (o.type == "ArrayPattern" || o.type == "ObjectPattern" || o.type == "AssignmentPattern" || o.type == "RestElement") {
          isDestructed = true;
          return "EXIT";
        }
      });

      if (isDestructed) {
        if (object.expression) {
          object.body = (0, _gen.BlockStatement)([(0, _gen.ReturnStatement)({ ...object.body
          })]);
        } else if (object.body.type != "BlockStatement") {
          object.body = (0, _gen.BlockStatement)([{ ...object.body
          }]);
        }

        var arrayPattern = (0, _gen.ArrayPattern)(parameters); // `arguments` is not allowed in arrow functions

        if (object.type == "ArrowFunctionExpression" && !object.params.find(x => x.type == "RestElement")) {
          // new names
          object.params = Array(object.params.length).fill(0).map(() => (0, _gen.Identifier)(this.getPlaceholder()));
          (0, _insert.getBlockBody)(object.body).unshift((0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(arrayPattern, (0, _gen.ArrayExpression)(object.params))));
        } else {
          object.params = [];
          (0, _insert.getBlockBody)(object.body).unshift((0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(arrayPattern, (0, _template.default)("Array.prototype.slice.call(arguments)").single().expression)));

          if (object.type == "ArrowFunctionExpression") {
            object.type = "FunctionExpression";
            object.expression = false;
            this.replace(object, (0, _gen.CallExpression)((0, _gen.MemberExpression)((0, _insert.clone)(object), (0, _gen.Identifier)("bind"), false), [(0, _gen.ThisExpression)()]));
          }
        }
      }
    };
  }

}
/**
 * Removes destructuring so the script can work in ES5 environments.
 */


class AntiDestructuring extends _transform.default {
  constructor(o) {
    super(o);
    this.before.push(new AntiDestructuringParameters(o));
  }

  match(object, parents) {
    return object.type == "AssignmentExpression" || object.type == "VariableDeclarator";
  }

  transform(object, parents) {
    var block = (0, _traverse.getBlock)(object, parents);
    var body = (0, _insert.getBlockBody)(block);
    var temp = this.getPlaceholder();
    var exprs = [];
    var names = new Set();
    var operator = "=";
    var id = null; // The object being set

    var extracting = null; // The object being extracted from

    if (object.type == "AssignmentExpression") {
      id = object.left;
      extracting = object.right;
      operator = object.operator;
    } else if (object.type == "VariableDeclarator") {
      id = object.id;
      extracting = object.init;
    } else {
      (0, _assert.ok)(false);
    }

    var should = false;
    (0, _traverse.walk)(id, [], (o, p) => {
      if (o.type && o.type.includes("Pattern")) {
        should = true;
      }
    });

    if (should) {
      (0, _insert.prepend)(block, (0, _gen.VariableDeclaration)([(0, _gen.VariableDeclarator)((0, _gen.Identifier)(temp))]));

      const recursive = (x, realm) => {
        realm = (0, _insert.clone)(realm);

        if (x.type == "Identifier") {
          exprs.push((0, _gen.AssignmentExpression)(operator, (0, _insert.clone)(x), realm));
          names.add(x.name);
        } else if (x.type == "ObjectPattern") {
          x.properties.forEach(property => {
            recursive(property.value, (0, _gen.MemberExpression)(realm, property.key, property.computed));
          });
        } else if (x.type == "ArrayPattern") {
          x.elements.forEach((element, i) => {
            if (element) {
              if (element.type == "RestElement") {
                if (i != x.elements.length - 1) {
                  this.error(new Error("Uncaught SyntaxError: Rest element must be last element"));
                }

                recursive(element.argument, (0, _gen.CallExpression)((0, _gen.MemberExpression)(realm, (0, _gen.Identifier)("slice"), false), [(0, _gen.Literal)(i)]));
              } else {
                recursive(element, (0, _gen.MemberExpression)(realm, (0, _gen.Literal)(i), true));
              }
            }
          });
        } else if (x.type == "AssignmentPattern") {
          var condition = (0, _gen.ConditionalExpression)((0, _gen.BinaryExpression)("===", realm, (0, _gen.Identifier)("undefined")), x.right, realm);
          recursive(x.left, condition);
        } else {
          throw new Error("unknown type: " + x.type);
        }
      };

      recursive(id, (0, _gen.Identifier)(temp));
      return () => {
        var seq = (0, _gen.SequenceExpression)([(0, _gen.AssignmentExpression)("=", (0, _gen.Identifier)(temp), (0, _insert.clone)(extracting) || (0, _gen.Identifier)("undefined")), ...exprs]);

        if (object.type == "VariableDeclarator") {
          var i = (0, _insert.getIndexDirect)(object, parents);
          var extra = Array.from(names).map(x => {
            return {
              type: "VariableDeclarator",
              id: (0, _gen.Identifier)(x),
              init: null
            };
          });
          extra.push({
            type: "VariableDeclarator",
            id: (0, _gen.Identifier)(this.getPlaceholder()),
            init: seq
          });
          parents[0].splice(i, 1, ...extra);
        } else {
          this.replace(object, seq);
        }
      };
    }
  }

}

exports.default = AntiDestructuring;
