// js-confuser/dist/transforms/calculator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transform = _interopRequireDefault(require("./transform"));

var _gen = require("../util/gen");

var _insert = require("../util/insert");

var _random = require("../util/random");

var _order = require("../order");

var _assert = require("assert");

var _precedence = require("../precedence");

var _template = _interopRequireDefault(require("../templates/template"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Calculator extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.Calculator);

    _defineProperty(this, "gen", void 0);

    _defineProperty(this, "ops", void 0);

    _defineProperty(this, "statesUsed", void 0);

    _defineProperty(this, "calculatorFn", void 0);

    _defineProperty(this, "calculatorOpVar", void 0);

    _defineProperty(this, "calculatorSetOpFn", void 0);

    this.ops = Object.create(null);
    this.statesUsed = new Set();
    this.calculatorFn = this.getPlaceholder();
    this.calculatorOpVar = this.getPlaceholder();
    this.calculatorSetOpFn = this.getPlaceholder();
    this.gen = this.getGenerator();
  }

  apply(tree) {
    super.apply(tree);

    if (Object.keys(this.ops).length == 0) {
      return;
    }

    var leftArg = this.getPlaceholder();
    var rightArg = this.getPlaceholder();
    var switchCases = [];
    Object.keys(this.ops).forEach(operator => {
      var code = this.ops[operator];
      var factory = operator == "&&" || operator == "||" ? _gen.LogicalExpression : _gen.BinaryExpression;
      var body = [(0, _gen.ReturnStatement)(factory(operator, (0, _gen.Identifier)(leftArg), (0, _gen.Identifier)(rightArg)))];
      switchCases.push((0, _gen.SwitchCase)((0, _gen.Literal)(code), body));
    });
    var func = (0, _gen.FunctionDeclaration)(this.calculatorFn, [leftArg, rightArg].map(x => (0, _gen.Identifier)(x)), [(0, _gen.SwitchStatement)((0, _gen.Identifier)(this.calculatorOpVar), switchCases)]);
    (0, _insert.prepend)(tree, (0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(this.calculatorOpVar)));
    (0, _insert.prepend)(tree, (0, _template.default)("function {name}(a){\n        a = {b} + ({b}=a, 0);\n        return a;\n      }").single({
      name: this.calculatorSetOpFn,
      b: this.calculatorOpVar
    }));
    (0, _insert.prepend)(tree, func);
  }

  match(object, parents) {
    return object.type == "BinaryExpression";
  }

  transform(object, parents) {
    var operator = object.operator;
    var allowedOperators = new Set(["+", "-", "*", "/"]);

    if (!allowedOperators.has(operator)) {
      return;
    }

    var myPrecedence = _precedence.OPERATOR_PRECEDENCE[operator] + Object.keys(_precedence.OPERATOR_PRECEDENCE).indexOf(operator) / 100;
    var precedences = parents.map(x => x.type == "BinaryExpression" && _precedence.OPERATOR_PRECEDENCE[x.operator] + Object.keys(_precedence.OPERATOR_PRECEDENCE).indexOf(x.operator) / 100); // corrupt AST

    if (precedences.find(x => x >= myPrecedence)) {
      return;
    }

    if (parents.find(x => x.$dispatcherSkip || x.type == "BinaryExpression")) {
      return;
    }

    return () => {
      if (typeof this.ops[operator] !== "number") {
        var newState;

        do {
          newState = (0, _random.getRandomInteger)(-1000, 1000 + Object.keys(this.ops).length * 5);
        } while (this.statesUsed.has(newState));

        (0, _assert.ok)(!isNaN(newState));
        this.statesUsed.add(newState);
        this.ops[operator] = newState;
        this.log(operator, "calc(".concat(newState, ", left, right)"));
      }

      this.replace(object, (0, _gen.CallExpression)((0, _gen.Identifier)(this.calculatorFn), [object.left, object.right, (0, _random.choice)([(0, _gen.AssignmentExpression)("=", (0, _gen.Identifier)(this.calculatorOpVar), (0, _gen.Literal)(this.ops[operator])), (0, _gen.CallExpression)((0, _gen.Identifier)(this.calculatorSetOpFn), [(0, _gen.Literal)(this.ops[operator])])])]));
    };
  }

}

exports.default = Calculator;
