// js-confuser/dist/transforms/opaquePredicates.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transform = _interopRequireDefault(require("./transform"));

var _gen = require("../util/gen");

var _random = require("../util/random");

var _order = require("../order");

var _insert = require("../util/insert");

var _template = _interopRequireDefault(require("../templates/template"));

var _probability = require("../probability");

var _assert = require("assert");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const testTypes = new Set(["ForStatement", "WhileStatement", "DoWhileStatement", "IfStatement", "ConditionExpression"]);

function isTestExpression(object, parents) {
  if (!object || !parents[0]) {
    return false;
  }

  if (testTypes.has(parents[0].type) && parents[0].test === object) {
    return true;
  }

  return false;
}
/**
 * Changes test expression (such as if statements, for loops) to add predicates.
 *
 * Predicates are computed at runtime.
 */


class OpaquePredicates extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.OpaquePredicates);

    _defineProperty(this, "undefinedVar", void 0);

    _defineProperty(this, "nullVar", void 0);

    _defineProperty(this, "numberVar", void 0);

    _defineProperty(this, "predicateName", void 0);

    _defineProperty(this, "predicate", void 0);

    _defineProperty(this, "predicates", void 0);

    _defineProperty(this, "gen", void 0);

    _defineProperty(this, "made", void 0);

    this.predicates = Object.create(null);
    this.gen = this.getGenerator((0, _random.getRandomInteger)(0, 20));
    this.made = 0;
  }

  match(object, parents) {
    return (isTestExpression(object, parents) || object.type == "SwitchCase") && !parents.find(x => x.$dispatcherSkip || x.type == "AwaitExpression");
  }

  transform(object, parents) {
    return () => {
      if (!(0, _probability.ComputeProbabilityMap)(this.options.opaquePredicates)) {
        return;
      }

      this.made++;

      if (this.made > 150) {
        return;
      }

      if (!this.predicate) {
        this.predicateName = this.getPlaceholder();
        this.predicate = (0, _gen.ObjectExpression)([]);
        var tempName = this.getPlaceholder();
        (0, _insert.prepend)(parents[parents.length - 1] || object, (0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(this.predicateName, (0, _gen.CallExpression)((0, _gen.FunctionExpression)([], [(0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(tempName, this.predicate)), (0, _gen.ReturnStatement)((0, _gen.Identifier)(tempName))]), []))));
      }

      var expr = (0, _random.choice)(Object.values(this.predicates));

      if (!expr || Math.random() < 0.5 / (Object.keys(this.predicates).length || 1)) {
        var prop = this.gen.generate();
        var accessor = (0, _gen.MemberExpression)((0, _gen.Identifier)(this.predicateName), (0, _gen.Identifier)(prop), false);

        switch ((0, _random.choice)(["array", "number", "string"])) {
          case "array":
            var arrayProp = this.gen.generate();
            this.predicate.properties.push((0, _gen.Property)((0, _gen.Identifier)(arrayProp), (0, _gen.ArrayExpression)([])));
            var paramName = this.getPlaceholder();
            this.predicate.properties.push((0, _gen.Property)((0, _gen.Identifier)(prop), (0, _gen.FunctionExpression)([(0, _gen.AssignmentPattern)((0, _gen.Identifier)(paramName), (0, _gen.Literal)("length"))], (0, _template.default)("\n                  if ( !".concat(this.predicateName, ".").concat(arrayProp, "[0] ) {\n                    ").concat(this.predicateName, ".").concat(arrayProp, ".push(").concat((0, _random.getRandomInteger)(-100, 100), ");\n                  }\n                  return ").concat(this.predicateName, ".").concat(arrayProp, "[").concat(paramName, "];\n                ")).compile())));
            expr = (0, _gen.CallExpression)(accessor, []);
            break;

          case "number":
            this.predicate.properties.push((0, _gen.Property)((0, _gen.Identifier)(prop), (0, _gen.Literal)((0, _random.getRandomInteger)(15, 90))));
            expr = (0, _gen.BinaryExpression)(">", accessor, (0, _gen.Literal)((0, _random.getRandomInteger)(-90, 10)));
            break;

          case "string":
            var str = this.gen.generate();
            var index = (0, _random.getRandomInteger)(0, str.length);
            var fn = Math.random() > 0.5 ? "charAt" : "charCodeAt";
            this.predicate.properties.push((0, _gen.Property)((0, _gen.Identifier)(prop), (0, _gen.Literal)(str)));
            expr = (0, _gen.BinaryExpression)("==", (0, _gen.CallExpression)((0, _gen.MemberExpression)(accessor, (0, _gen.Literal)(fn), true), [(0, _gen.Literal)(index)]), (0, _gen.Literal)(str[fn](index)));
            break;
        }

        (0, _assert.ok)(expr);
        this.predicates[prop] = expr;

        if (Math.random() > 0.8) {
          (0, _random.shuffle)(this.predicate.properties);
        }
      }

      var cloned = (0, _insert.clone)(expr);

      if (object.type == "SwitchCase") {
        var matching = (0, _gen.Identifier)((0, _random.choice)(["undefined", "null"]));
        var test = object.test;

        if (test.type == "Literal") {
          if (typeof test.value === "number") {
            matching = (0, _gen.Literal)((0, _random.getRandomInteger)(-250, 250));
          } else if (typeof test.value === "string") {
            matching = (0, _gen.Literal)((0, _random.getRandomString)(4));
          }
        }

        var conditionalExpression = (0, _gen.ConditionalExpression)(cloned, (0, _insert.clone)(test), matching);

        if (Math.random() > 0.5) {
          conditionalExpression = (0, _gen.ConditionalExpression)((0, _gen.UnaryExpression)("!", cloned), matching, (0, _insert.clone)(test));
        }

        this.replace(test, conditionalExpression);
      } else if (isTestExpression(object, parents)) {
        if (object.type == "Literal" && !object.regex) {
          if (object.value) {
            this.replace(object, cloned);
          } else {
            this.replace(object, (0, _gen.UnaryExpression)("!", cloned));
          }
        } else {
          this.replace(object, (0, _gen.LogicalExpression)("&&", (0, _insert.clone)(object), cloned));
        }
      }
    };
  }

}

exports.default = OpaquePredicates;
