// js-confuser/dist/transforms/renameLabels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _order = require("../order");

var _traverse = require("../traverse");

var _compare = require("../util/compare");

var _gen = require("../util/gen");

var _insert = require("../util/insert");

var _transform = _interopRequireDefault(require("./transform"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Renames the labels to shorter names.
 */
class RenameLabels extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.RenameLabels);

    _defineProperty(this, "gen", void 0);

    this.gen = this.getGenerator();
  }

  match(object, parents) {
    return object.type == "LabeledStatement";
  }

  transform(object, parents) {
    return () => {
      var newName = null;
      var isRemovable = object.body.type !== "BlockStatement";
      var labelNeverUsed = true;
      (0, _traverse.walk)(object, parents, (o, p) => {
        if (o.type == "BreakStatement" || o.type == "ContinueStatement") {
          function isContinuableStatement(x, stmtParents) {
            return (0, _compare.isLoop)(x) && x.type !== "SwitchStatement";
          }

          function isBreakableStatement(x, stmtParents) {
            return (0, _compare.isLoop)(x) || x.type == "BlockStatement" && o.label && stmtParents[0] && stmtParents[0].type == "LabeledStatement";
          }

          var fn = o.type == "ContinueStatement" ? isContinuableStatement : isBreakableStatement;
          var labelStatement = p.find((node, i) => {
            return fn(node, p.slice(i + 1));
          });

          if (o.label && o.label.name == object.label.name) {
            if (object.body == labelStatement && isRemovable) {
              // In same loop
              o.label = null;
            } else {
              if (!newName) {
                newName = this.gen.generate();
              }

              o.label = (0, _gen.Identifier)(newName);
              labelNeverUsed = false;
            }
          }
        }
      });

      if (newName) {
        object.label = (0, _gen.Identifier)(newName);
      } else if (isRemovable || labelNeverUsed) {
        this.replace(object, (0, _insert.clone)(object.body));
      }
    };
  }

}

exports.default = RenameLabels;
