// js-confuser/dist/transforms/identifier/globalConcealing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _template = _interopRequireDefault(require("../../templates/template"));

var _transform = _interopRequireDefault(require("../transform"));

var _order = require("../../order");

var _gen = require("../../util/gen");

var _insert = require("../../util/insert");

var _identifiers = require("../../util/identifiers");

var _random = require("../../util/random");

var _constants = require("../../constants");

var _probability = require("../../probability");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class GlobalAnalysis extends _transform.default {
  constructor(o) {
    super(o);

    _defineProperty(this, "notGlobals", void 0);

    _defineProperty(this, "globals", void 0);

    this.globals = Object.create(null);
    this.notGlobals = new Set();
  }

  match(object, parents) {
    return object.type == "Identifier" && !_constants.reservedKeywords.has(object.name);
  }

  transform(object, parents) {
    // no touching `import()` or `import x from ...`
    var importIndex = parents.findIndex(x => x.type == "ImportExpression" || x.type == "ImportDeclaration");

    if (importIndex !== -1) {
      if (parents[importIndex].source === (parents[importIndex - 1] || object)) {
        return;
      }
    }

    var info = (0, _identifiers.getIdentifierInfo)(object, parents);

    if (!info.spec.isReferenced) {
      return;
    } // Add to globals


    if (!this.notGlobals.has(object.name)) {
      if (!this.globals[object.name]) {
        this.globals[object.name] = [];
      }

      this.globals[object.name].push([object, parents]);
    }

    if (info.spec.isDefined || info.spec.isModified) {
      delete this.globals[object.name];
      this.notGlobals.add(object.name);
    }

    var assignmentIndex = parents.findIndex(x => x.type == "AssignmentExpression");
    var updateIndex = parents.findIndex(x => x.type == "UpdateExpression");

    if (assignmentIndex != -1 && parents[assignmentIndex].left === (parents[assignmentIndex - 1] || object) || updateIndex != -1) {
      var memberIndex = parents.findIndex(x => x.type == "MemberExpression");

      if (memberIndex == -1 || memberIndex > (assignmentIndex == -1 ? assignmentIndex : updateIndex)) {
        delete this.globals[object.name];
        this.notGlobals.add(object.name);
      }
    }
  }

}
/**
 * Global Concealing hides global variables being accessed.
 *
 * - Any variable that is not defined is considered "global"
 */


class GlobalConcealing extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.GlobalConcealing);

    _defineProperty(this, "globalAnalysis", void 0);

    this.globalAnalysis = new GlobalAnalysis(o);
    this.before.push(this.globalAnalysis);
  }

  match(object, parents) {
    return object.type == "Program";
  }

  transform(object, parents) {
    return () => {
      var globals = this.globalAnalysis.globals;
      this.globalAnalysis.notGlobals.forEach(del => {
        delete globals[del];
      });
      delete globals["require"];

      _constants.reservedIdentifiers.forEach(x => {
        delete globals[x];
      });

      Object.keys(globals).forEach(x => {
        if (this.globalAnalysis.globals[x].length < 1) {
          delete globals[x];
        } else if (!(0, _probability.ComputeProbabilityMap)(this.options.globalConcealing, x => x, x)) {
          delete globals[x];
        }
      }); // this.log(Object.keys(globals).join(', '))

      if (Object.keys(globals).length > 0) {
        var used = new Set(); // 1. Make getter function
        // holds "window" or "global"

        var globalVar = this.getPlaceholder(); // holds outermost "this"

        var thisVar = this.getPlaceholder(); // "window" or "global" in node

        var global = this.options.globalVariables.values().next().value || "window";
        var getGlobalVariableFnName = this.getPlaceholder();
        var getThisVariableFnName = this.getPlaceholder(); // Returns global variable or fall backs to `this`

        var getGlobalVariableFn = (0, _template.default)("\n        var ".concat(getGlobalVariableFnName, " = function(){\n          try {\n            return ").concat(global, ";\n          } catch (e){\n            return ").concat(getThisVariableFnName, "[\"call\"](this);\n          }\n        }")).single();
        var getThisVariableFn = (0, _template.default)("\n        var ".concat(getThisVariableFnName, " = function(){\n          try {\n            return this;\n          } catch (e){\n            return null;\n          }\n        }")).single(); // 2. Replace old accessors

        var globalFn = this.getPlaceholder();
        var newNames = Object.create(null);
        Object.keys(globals).forEach(name => {
          var locations = globals[name];
          var state;

          do {
            state = (0, _random.getRandomInteger)(-1000, 1000 + used.size);
          } while (used.has(state));

          used.add(state);
          newNames[name] = state;
          locations.forEach(_ref => {
            let [node, parents] = _ref;

            if (!parents.find(x => x.$dispatcherSkip)) {
              // Do not replace
              if (parents[0]) {
                if (parents[0].type == "ClassDeclaration" || parents[0].type == "ClassExpression" || parents[0].type == "FunctionExpression" || parents[0].type == "FunctionDeclaration") {
                  if (parents[0].id === node) {
                    return;
                  }
                }
              }

              this.replace(node, (0, _gen.CallExpression)((0, _gen.Identifier)(globalFn), [(0, _gen.Literal)(state)]));
            }
          });
        }); // Adds all global variables to the switch statement

        this.options.globalVariables.forEach(name => {
          if (!newNames[name]) {
            var state;

            do {
              state = (0, _random.getRandomInteger)(-1000, 1000 + used.size + this.options.globalVariables.size * 100);
            } while (used.has(state));

            used.add(state);
            newNames[name] = state;
          }
        });
        var indexParamName = this.getPlaceholder();
        var returnName = this.getPlaceholder();
        var functionDeclaration = (0, _gen.FunctionDeclaration)(globalFn, [(0, _gen.Identifier)(indexParamName)], [(0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(returnName)), (0, _gen.SwitchStatement)((0, _gen.Identifier)(indexParamName), Object.keys(newNames).map(name => {
          var code = newNames[name];
          var body = [(0, _gen.ReturnStatement)((0, _gen.LogicalExpression)("||", (0, _gen.MemberExpression)((0, _gen.Identifier)(globalVar), (0, _gen.Literal)(name), true), (0, _gen.MemberExpression)((0, _gen.Identifier)(thisVar), (0, _gen.Literal)(name), true)))];

          if (Math.random() > 0.5 && name) {
            body = [(0, _gen.ExpressionStatement)((0, _gen.AssignmentExpression)("=", (0, _gen.Identifier)(returnName), (0, _gen.LogicalExpression)("||", (0, _gen.Literal)(name), (0, _gen.MemberExpression)((0, _gen.Identifier)(thisVar), (0, _gen.Literal)(name), true)))), (0, _gen.BreakStatement)()];
          }

          return (0, _gen.SwitchCase)((0, _gen.Literal)(code), body);
        })), (0, _gen.ReturnStatement)((0, _gen.LogicalExpression)("||", (0, _gen.MemberExpression)((0, _gen.Identifier)(globalVar), (0, _gen.Identifier)(returnName), true), (0, _gen.MemberExpression)((0, _gen.Identifier)(thisVar), (0, _gen.Identifier)(returnName), true)))]);
        var tempVar = this.getPlaceholder();
        var variableDeclaration = (0, _template.default)("\n        var ".concat(globalVar, ", ").concat(thisVar, ";\n        ")).single();
        variableDeclaration.declarations.push((0, _gen.VariableDeclarator)(tempVar, (0, _gen.CallExpression)((0, _gen.MemberExpression)((0, _gen.FunctionExpression)([], [getGlobalVariableFn, getThisVariableFn, (0, _template.default)("return ".concat(thisVar, " = ").concat(getThisVariableFnName, "[\"call\"](this, ").concat(globalFn, "), ").concat(globalVar, " = ").concat(getGlobalVariableFnName, "[\"call\"](this)")).single()]), (0, _gen.Literal)("call"), true), [])));
        (0, _insert.prepend)(object, variableDeclaration);
        (0, _insert.append)(object, functionDeclaration);
      }
    };
  }

}

exports.default = GlobalConcealing;
