// js-confuser/dist/transforms/identifier/nameRecycling.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = require("assert");

var _order = require("../../order");

var _probability = require("../../probability");

var _traverse = require("../../traverse");

var _gen = require("../../util/gen");

var _identifiers = require("../../util/identifiers");

var _insert = require("../../util/insert");

var _transform = _interopRequireDefault(require("../transform"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Statement-based variable recycling.
 *
 * ```js
 * // Input
 * function percentage(decimal) {
 *   var multiplied = x * 100;
 *   var floored = Math.floor(multiplied);
 *   var output = floored + "%"
 *   return output;
 * }
 *
 * // Output
 * function percentage(decimal) {
 *   var multiplied = x * 100;
 *   var floored = Math.floor(multiplied);
 *   multiplied = floored + "%";
 *   return multiplied;
 * }
 * ```
 */
class NameRecycling extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.NameRecycling);
  }

  match(object, parents) {
    return (0, _traverse.isBlock)(object);
  }

  transform(object, parents) {
    return () => {
      if ((0, _identifiers.containsLexicallyBoundVariables)(object, parents)) {
        return;
      }

      var context = (0, _insert.getVarContext)(object, parents);
      var stmts = [...object.body];
      (0, _assert.ok)(Array.isArray(stmts));
      var definedMap = new Map();
      var referencedMap = new Map();
      var nodeMap = new Map();
      var lastReferenceMap = new Map();
      var defined = new Set();
      var illegal = new Set();
      var fn = (0, _insert.isFunction)(parents[0]) ? parents[0] : null;

      if (fn) {
        definedMap.set(-1, new Set((0, _identifiers.getFunctionParameters)(fn, parents.slice(1)).map(x => x[0].name)));
      }

      stmts.forEach((stmt, i) => {
        var definedHere = new Set();
        var referencedHere = new Set();
        var nodesHere = new Map();
        (0, _traverse.walk)(stmt, [object.body, object, ...parents], (o, p) => {
          if (o.type == "Identifier") {
            return () => {
              var info = (0, _identifiers.getIdentifierInfo)(o, p);

              if (!info.spec.isReferenced) {
                return;
              }

              lastReferenceMap.set(o.name, i);
              var comparingContext = info.spec.isDefined ? (0, _insert.getDefiningContext)(o, p) : (0, _insert.getReferencingContexts)(o, p).find(x => (0, _insert.isVarContext)(x));

              if (comparingContext !== context) {
                illegal.add(o.name);
                this.log(o.name, "is different context");
              } else {
                if (!nodesHere.has(o.name)) {
                  nodesHere.set(o.name, [[o, p]]);
                } else {
                  nodesHere.get(o.name).push([o, p]);
                }

                if (info.spec.isDefined) {
                  if (defined.has(o.name) || (0, _traverse.getBlock)(o, p) !== object) {
                    illegal.add(o.name);
                  }

                  defined.add(o.name);
                  definedHere.add(o.name);
                } else {
                  referencedHere.add(o.name);
                }
              }
            };
          }
        }); // console.log(i, definedHere);

        definedMap.set(i, definedHere);
        referencedMap.set(i, referencedHere);
        nodeMap.set(i, nodesHere);
      });
      this.log(illegal);
      illegal.forEach(name => {
        nodeMap.forEach(value => {
          value.delete(name);
        });
      });
      var available = new Set();
      var newNames = Object.create(null);
      stmts.forEach((stmt, i) => {
        var nodes = nodeMap.get(i);
        nodes.forEach((locations, name) => {
          var newName = newNames[name];

          if (!newName) {
            var canChange = false;

            if (object.type == "Program" && !(0, _probability.ComputeProbabilityMap)(this.options.renameGlobals, x => x, name)) {
              return;
            }

            if (defined.has(name) && definedMap.get(i).has(name)) {
              canChange = true;
            }

            if (!canChange) {
              return;
            }

            if (available.size) {
              newName = available.keys().next().value;
              available.delete(newName);
              (0, _assert.ok)(name !== newName);
              newNames[name] = newName;
              defined.delete(name);
              this.log(name, "->", newName);
            }
          }
        });
        nodes.forEach((locations, name) => {
          var newName = newNames[name];

          if (newName) {
            locations.forEach(_ref => {
              let [object, parents] = _ref;
              object.name = newName;
              var declaratorIndex = parents.findIndex(p => p.type == "VariableDeclarator");

              if (declaratorIndex !== -1 && parents[declaratorIndex].id === (parents[declaratorIndex - 1] || object)) {
                var value = parents[declaratorIndex].init || (0, _gen.Identifier)("undefined");
                var expr = (0, _gen.AssignmentExpression)("=", parents[declaratorIndex].id, value);

                if (parents[declaratorIndex + 1].length === 1) {
                  if ((0, _insert.isForInitialize)(parents[declaratorIndex + 2], parents.slice(3))) {
                    this.replace(parents[declaratorIndex + 2], expr);
                  } else {
                    this.replace(parents[declaratorIndex + 2], (0, _gen.ExpressionStatement)(expr));
                  }
                } else {
                  this.replace(parents[declaratorIndex], (0, _gen.VariableDeclarator)(this.getPlaceholder(), expr));
                }
              } else {
                if (parents[0].type == "FunctionDeclaration") {
                  this.replace(parents[0], (0, _gen.ExpressionStatement)((0, _gen.AssignmentExpression)("=", (0, _gen.Identifier)(newName), { ...parents[0],
                    type: "FunctionExpression",
                    id: null
                  })));
                } else if (parents[0].type == "ClassDeclaration") {
                  this.replace(parents[0], (0, _gen.ExpressionStatement)((0, _gen.AssignmentExpression)("=", (0, _gen.Identifier)(newName), { ...parents[0],
                    type: "ClassExpression"
                  })));
                }
              }
            });
          }

          if (defined.has(name)) {
            var lastRef = lastReferenceMap.get(name);
            var isDecommissioned = lastRef === i;

            if (isDecommissioned) {
              available.add(name);
            }
          }
        });
      });
    };
  }

}

exports.default = NameRecycling;
