// js-confuser/dist/transforms/preparation/preparation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transform = _interopRequireDefault(require("../transform"));

var _gen = require("../../util/gen");

var _order = require("../../order");

var _insert = require("../../util/insert");

var _identifiers = require("../../util/identifiers");

var _label = _interopRequireDefault(require("../label"));

var _antiDestructuring = _interopRequireDefault(require("../es5/antiDestructuring"));

var _compare = require("../../util/compare");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * The file contains all preparation transformations
 */

/**
 * People use shortcuts and its harder to parse.
 *
 * - `if (a) b()` -> `if (a) { b() }`
 * - Ensures all bodies are `BlockStatement`, not individual expression statements
 */
class Block extends _transform.default {
  constructor(o) {
    super(o);
  }

  match(object, parents) {
    return !Array.isArray(object);
  }

  transform(object, parents) {
    switch (object.type) {
      case "IfStatement":
        if (object.consequent.type != "BlockStatement") {
          object.consequent = (0, _gen.BlockStatement)([(0, _insert.clone)(object.consequent)]);
        }

        if (object.alternate && object.alternate.type != "BlockStatement") {
          object.alternate = (0, _gen.BlockStatement)([(0, _insert.clone)(object.alternate)]);
        }

        break;

      case "WhileStatement":
      case "WithStatement":
      case "ForStatement":
      case "ForOfStatement":
      case "ForInStatement":
        if (object.body.type != "BlockStatement") {
          object.body = (0, _gen.BlockStatement)([(0, _insert.clone)(object.body)]);
        }

        break;

      case "ArrowFunctionExpression":
        if (object.body.type !== "BlockStatement" && object.expression) {
          object.body = (0, _gen.BlockStatement)([(0, _gen.ReturnStatement)((0, _insert.clone)(object.body))]);
          object.expression = false;
        }

        break;
    }
  }

}

class ExplicitIdentifiers extends _transform.default {
  constructor(o) {
    super(o);
  }

  match(object, parents) {
    return object.type == "Identifier";
  }

  transform(object, parents) {
    if (object.name === "eval") {
      var fn = (0, _insert.getFunction)(object, parents);

      if (fn) {
        fn.$requiresEval = true;
      }
    }

    var info = (0, _identifiers.getIdentifierInfo)(object, parents);

    if (info.isPropertyKey || info.isAccessor) {
      var propIndex = parents.findIndex(x => x.type == "MethodDefinition" || x.type == "Property");

      if (propIndex !== -1) {
        if (parents[propIndex].type == "MethodDefinition" && parents[propIndex].kind == "constructor") {
          return;
        }
      }

      this.log(object.name, "->", "'".concat(object.name, "'"));
      this.replace(object, (0, _gen.Literal)(object.name));
      parents[0].computed = true;
      parents[0].shorthand = false;
    }
  }

}

class ExplicitDeclarations extends _transform.default {
  constructor(o) {
    super(o);
  }

  match(object, parents) {
    return object.type == "VariableDeclaration";
  }

  transform(object, parents) {
    // for ( var x in ... ) {...}
    var forIndex = parents.findIndex(x => x.type == "ForInStatement" || x.type == "ForOfStatement");

    if (forIndex != -1 && parents[forIndex].left == (parents[forIndex - 1] || object)) {
      object.declarations.forEach(x => {
        x.init = null;
      });
      return;
    }

    var body = parents[0];

    if ((0, _compare.isLoop)(body) || body.type == "LabeledStatement") {
      return;
    }

    if (body.type == "ExportNamedDeclaration") {
      return;
    }

    if (!Array.isArray(body)) {
      this.error(new Error("body is " + body.type));
    }

    if (object.declarations.length > 1) {
      // Make singular
      var index = body.indexOf(object);

      if (index == -1) {
        this.error(new Error("index is -1"));
      }

      var after = object.declarations.slice(1);
      body.splice(index + 1, 0, ...after.map(x => {
        return {
          type: "VariableDeclaration",
          declarations: [(0, _insert.clone)(x)],
          kind: object.kind
        };
      }));
      object.declarations.length = 1;
    }
  }

}

class Preparation extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.Preparation);
    this.before.push(new Block(o));
    this.before.push(new _label.default(o));
    this.before.push(new ExplicitIdentifiers(o));
    this.before.push(new ExplicitDeclarations(o));

    if (this.options.es5) {
      this.before.push(new _antiDestructuring.default(o));
    } // this.before.push(new NameConflicts(o));

  }

  match() {
    return false;
  }

}

exports.default = Preparation;
