// js-confuser/dist/transforms/string/stringCompression.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = require("assert");

var _order = require("../../order");

var _probability = require("../../probability");

var _template = _interopRequireDefault(require("../../templates/template"));

var _compare = require("../../util/compare");

var _gen = require("../../util/gen");

var _insert = require("../../util/insert");

var _transform = _interopRequireDefault(require("../transform"));

var _stringConcealing = require("./stringConcealing");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function LZ_encode(c) {
  (0, _assert.ok)(c);
  var x = "charCodeAt",
      b,
      e = {},
      f = c.split(""),
      d = [],
      a = f[0],
      g = 256;

  for (b = 1; b < f.length; b++) c = f[b], null != e[a + c] ? a += c : (d.push(1 < a.length ? e[a] : a[x](0)), e[a + c] = g, g++, a = c);

  d.push(1 < a.length ? e[a] : a[x](0));

  for (b = 0; b < d.length; b++) d[b] = String.fromCharCode(d[b]);

  return d.join("");
}

function LZ_decode(b) {
  (0, _assert.ok)(b);
  var o,
      f,
      a,
      e = {},
      d = b.split(""),
      c = f = d[0],
      g = [c],
      h = o = 256;

  for (var i = 1; i < d.length; i++) a = d[i].charCodeAt(0), a = h > a ? d[i] : e[a] ? e[a] : f + c, g.push(a), c = a.charAt(0), e[o] = f + c, o++, f = a;

  return g.join("");
}

const DecodeTemplate = (0, _template.default)("function {name}(b){\n    var o,\n    f,\n    a,\n    e = {},\n    d = b.split(\"\"),\n    c = (f = d[0]),\n    g = [c],\n    h = (o = 256);\n  for (b = 1; b < d.length; b++)\n    (a = d[b].charCodeAt(0)),\n      (a = h > a ? d[b] : e[a] ? e[a] : f + c),\n      g.push(a),\n      (c = a.charAt(0)),\n      (e[o] = f + c),\n      o++,\n      (f = a);\n  return g.join(\"\").split(\"{delimiter}\");\n  }");

class StringCompression extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.StringCompression);

    _defineProperty(this, "map", void 0);

    _defineProperty(this, "ignore", void 0);

    _defineProperty(this, "string", void 0);

    _defineProperty(this, "delimiter", "1");

    _defineProperty(this, "fnName", void 0);

    this.map = new Map();
    this.ignore = new Set();
    this.string = "";
    this.fnName = this.getPlaceholder();
  }

  apply(tree) {
    super.apply(tree);
    this.string = this.string.slice(0, this.string.length - 1);

    if (!this.string.length) {
      return;
    }

    var split = this.getPlaceholder();
    var decoder = this.getPlaceholder();
    var getStringName = this.getPlaceholder();
    var encoded = LZ_encode(this.string);

    if (LZ_decode(encoded) !== this.string) {
      this.error(new Error("String failed to be decoded"));
    }

    var getStringParamName = this.getPlaceholder();
    var decoderParamName = this.getPlaceholder();
    var callExpression = (0, _gen.CallExpression)((0, _gen.Identifier)(decoderParamName), [(0, _gen.CallExpression)((0, _gen.Identifier)(getStringParamName), [])]);
    (0, _insert.prepend)(tree, (0, _gen.VariableDeclaration)((0, _gen.VariableDeclarator)(split, (0, _gen.CallExpression)((0, _gen.FunctionExpression)([(0, _gen.Identifier)(getStringParamName), (0, _gen.Identifier)(decoderParamName)], [(0, _gen.ReturnStatement)(callExpression)]), [(0, _gen.Identifier)(getStringName), (0, _gen.Identifier)(decoder)]))));
    (0, _insert.append)(tree, (0, _gen.FunctionDeclaration)(getStringName, [], [(0, _gen.ReturnStatement)((0, _gen.Literal)(encoded))]));
    (0, _insert.append)(tree, (0, _gen.FunctionDeclaration)(this.fnName, [(0, _gen.Identifier)("index")], [(0, _gen.ReturnStatement)((0, _gen.MemberExpression)((0, _gen.Identifier)(split), (0, _gen.Identifier)("index"), true))]));
    (0, _insert.append)(tree, DecodeTemplate.single({
      name: decoder,
      delimiter: this.delimiter
    }));
  }

  match(object, parents) {
    return object.type == "Literal" && typeof object.value === "string" && object.value && object.value.length > 3 && !(0, _compare.isDirective)(object, parents) && !(0, _stringConcealing.isModuleSource)(object, parents);
  }

  transform(object, parents) {
    if (!object.value) {
      return;
    }

    if (this.ignore.has(object.value) || object.value.includes(this.delimiter)) {
      return;
    }

    if (!parents[0] || parents[0].type == "CallExpression" && parents[0].callee.type == "Identifier" && parents[0].callee.name == this.fnName) {
      return;
    }

    if (!(0, _probability.ComputeProbabilityMap)(object.value, x => x)) {
      return;
    }

    var index = this.map.get(object.value);

    if (typeof index !== "number") {
      if (LZ_decode(LZ_encode(object.value)) !== object.value) {
        this.ignore.add(object.value);
        return;
      }

      var before = this.string;
      index = this.map.size;
      this.map.set(object.value, index);
      this.string += object.value + this.delimiter; // allow rollback if string becomes corrupted

      if (LZ_decode(LZ_encode(this.string)) !== this.string) {
        this.string = before;
        this.map.delete(object.value);
        this.ignore.add(object.value);
        return;
      }
    }

    (0, _assert.ok)(typeof index === "number");
    this.replaceIdentifierOrLiteral(object, (0, _gen.CallExpression)((0, _gen.Identifier)(this.fnName), [(0, _gen.Literal)(index)]), parents);
  }

}

exports.default = StringCompression;
