// js-confuser/dist/transforms/string/stringSplitting.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transform = _interopRequireDefault(require("../transform"));

var _gen = require("../../util/gen");

var _insert = require("../../util/insert");

var _random = require("../../util/random");

var _order = require("../../order");

var _stringConcealing = require("./stringConcealing");

var _compare = require("../../util/compare");

var _assert = require("assert");

var _probability = require("../../probability");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class StringSplitting extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.StringSplitting);

    _defineProperty(this, "joinPrototype", void 0);

    _defineProperty(this, "strings", void 0);

    _defineProperty(this, "adders", void 0);

    _defineProperty(this, "vars", void 0);

    this.joinPrototype = null;
    this.strings = Object.create(null);
    this.adders = [];
    this.vars = [];
  }

  apply(tree) {
    super.apply(tree);

    if (this.vars.length) {
      (0, _random.shuffle)(this.adders);
      (0, _random.shuffle)(this.vars);
      var body = tree.body;
      this.adders.forEach(nodes => {
        nodes.forEach(x => body.unshift(x));
      });
      var variableDeclaration = {
        type: "VariableDeclaration",
        declarations: [],
        kind: "var"
      };
      this.vars.forEach(node => variableDeclaration.declarations.push(node));
      body.unshift(variableDeclaration);
    }
  }

  match(object, parents) {
    return object.type == "Literal" && typeof object.value === "string" && !(0, _stringConcealing.isModuleSource)(object, parents) && !(0, _compare.isDirective)(object, parents);
  }

  transform(object, parents) {
    return () => {
      var size = Math.round(Math.max(6, object.value.length / (0, _random.getRandomInteger)(3, 8)));

      if (object.value.length <= size) {
        return;
      }

      var chunks = (0, _random.splitIntoChunks)(object.value, size);

      if (!chunks || chunks.length <= 1) {
        return;
      }

      if (!(0, _probability.ComputeProbabilityMap)(this.options.stringSplitting, x => x, object.value)) {
        return;
      }

      var binaryExpression;
      var parent;
      var last = chunks.pop();
      chunks.forEach((chunk, i) => {
        if (i == 0) {
          (0, _assert.ok)(i == 0);
          parent = binaryExpression = (0, _gen.BinaryExpression)("+", (0, _gen.Literal)(chunk), (0, _gen.Literal)(""));
        } else {
          binaryExpression.left = (0, _gen.BinaryExpression)("+", (0, _insert.clone)(binaryExpression.left), (0, _gen.Literal)(chunk));
          (0, _assert.ok)(binaryExpression);
        }
      });
      parent.right = (0, _gen.Literal)(last);
      this.replaceIdentifierOrLiteral(object, parent, parents);
    };
  }

}

exports.default = StringSplitting;
