// js-confuser/dist/transforms/string/stringEncoding.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transform = _interopRequireDefault(require("../transform"));

var _random = require("../../util/random");

var _order = require("../../order");

var _compare = require("../../util/compare");

var _stringConcealing = require("./stringConcealing");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function pad(x, len) {
  while (x.length < len) {
    x = "0" + x;
  }

  return x;
}

function even(x) {
  if (x.length % 2 != 0) {
    return "0" + x;
  }

  return x;
}

function toHexRepresentation(str) {
  var escapedString = "";
  str.split("").forEach(char => {
    var code = char.charCodeAt(0);

    if (code < 128) {
      escapedString += "\\x" + even(pad(code.toString(16), 2));
    } else {
      escapedString += char;
    }
  });
  return escapedString;
}

function toUnicodeRepresentation(str) {
  var escapedString = "";
  str.split("").forEach(char => {
    var code = char.charCodeAt(0);

    if (code < 128) {
      escapedString += "\\u" + even(pad(code.toString(16), 4));
    } else {
      escapedString += char;
    }
  });
  return escapedString;
}
/**
 * [String Encoding](https://docs.jscrambler.com/code-integrity/documentation/transformations/string-encoding) transforms a string into an encoded representation.
 *
 * - Potency Low
 * - Resilience Low
 * - Cost Low
 */


class StringEncoding extends _transform.default {
  constructor(o) {
    super(o, _order.ObfuscateOrder.StringEncoding);

    _defineProperty(this, "seen", void 0);
  }

  apply(tree) {
    this.seen = new Set();
    super.apply(tree);
  }

  match(object, parents) {
    return object.type == "Literal" && typeof object.value === "string" && !(0, _stringConcealing.isModuleSource)(object, parents) && !(0, _compare.isDirective)(object, parents);
  }

  transform(object, parents) {
    // Todo fix circular json problems
    if (this.seen.has(object)) {
      return;
    }

    this.seen.add(object);
    var type = (0, _random.choice)(["hexadecimal", "unicode"]);
    var escapedString = (type == "hexadecimal" ? toHexRepresentation : toUnicodeRepresentation)(object.value); // escodegen tries to escape backslashes, here is a work-around

    this.replace(object, {
      type: "Identifier",
      name: "'".concat(escapedString, "'")
    });
  }

}

exports.default = StringEncoding;
